File: iosApp/iosApp/Screens/Login/LoginView.swift
//
//  LoginView.swift
//  iosApp
//
//  Created by Дмитрий Окунев on 08.11.2022.
//  Copyright © 2022 orgName. All rights reserved.
//

import SwiftUI
import SharedSDK


struct LoginView: View {
    
    let viewState: LoginViewState
    let eventHandler: (LoginEvent) -> Void
    
    var body: some View {
        VStack {
            VStack{
                Spacer().frame(height: 36)
                Text("Login Now")
                    .foregroundColor(.textPrimary)
                    .fontWeight(.bold)
                    .font(.system(size: 24))
                Text("Welcome back to PlayZone! Enter your email address and your password to enjoy the latest features of PlayZone")
                    .foregroundColor(.textPrimary.opacity(0.5))
                    .fixedSize(horizontal: false, vertical: true)
                    .multilineTextAlignment(.center)
                    .padding(EdgeInsets(top: 16, leading: 30, bottom: 0, trailing: 30))
                
                Spacer().frame(height: 50)
                
                CommonTextField(hint: "Email Address", enabled: !viewState.isSending) {newValue in
                    eventHandler(.EmailChanged(value: newValue))
                }
                Spacer().frame(height: 24)
                CommonTextField(hint: "Password", enabled: !viewState.isSending, isSecure: !viewState.passwordHidden) { newValue in
                    eventHandler(.PasswordChanged(value: newValue))
                }
                LoginActionView (viewState: viewState ,onForgotClicked: {
                    eventHandler(.ForgotClick())
                }, onSubmitClicked: {
                    eventHandler(.LoginClick())
                })
            }
            
            Spacer()
            
            
            HStack{
                Text("Don't have account?")
                    .foregroundColor(.textPrimary)
                    .opacity(0.5)
                Text("Create one")
                    .foregroundColor(.textTintColorPrimary)
                    .fontWeight(.bold)
                    .onTapGesture {
                        eventHandler(.RegistrationClick())
                    }
            }
        }
    }
}

struct LoginActionView: View {
    let viewState: LoginViewState
    let onForgotClicked: () -> Void
    let onSubmitClicked: () -> Void
    
    var body: some View {
        VStack {
            Spacer().frame(height: 30)
            
            HStack {
                Spacer()
                Text("Forgot password")
                    .foregroundColor(.textTintColorPrimary)
                    .onTapGesture {
                        onForgotClicked()
                    }
                Spacer().frame(width: 30)
            }
            Spacer().frame(height: 30)
            ActionButton(title: "Login Now", enabled: !viewState.isSending) {
                onSubmitClicked()
            }
        }
    }
}

struct LoginView_Previews: PreviewProvider {
    static var previews: some View {
        LoginView(viewState: LoginViewState(email: "okdiman@rambler.ru", password: "12ed1c1c", passwordHidden: true, isSending: false), eventHandler: { event in
            
        }).background(Color.backgroundPrimary)
            .background(ignoresSafeAreaEdges: [.top, .bottom])
    }
}

File: common/auth/compose/src/commonMain/kotlin/login/LoginView.kt
package login

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material.Button
import androidx.compose.material.ButtonDefaults
import androidx.compose.material.Icon
import androidx.compose.material.Text
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Clear
import androidx.compose.material.icons.outlined.Lock
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import login.models.LoginEvent
import login.models.LoginViewState
import theme.Theme
import widgets.CommonTextField

@Composable
fun LoginView(viewState: LoginViewState, eventHandler: (LoginEvent) -> Unit) {
    Column(
        modifier = Modifier.padding(top = 36.dp, start = 24.dp, end = 24.dp, bottom = 50.dp),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        /**
         *  не выношу строки в ресурсы, так как такой подход не будет работать с другими платформами
         *  кроме Андроид
         *  P.S. разобраться с либами для работы с ресурсами, например mockResources
         */
        /**
         *  не выношу строки в ресурсы, так как такой подход не будет работать с другими платформами
         *  кроме Андроид
         *  P.S. разобраться с либами для работы с ресурсами, например mockResources
         */
        Text(
            text = "Login Now",
            color = Theme.colors.secondaryTextColor,
            fontSize = 24.sp,
            fontWeight = FontWeight.Bold
        )
        Spacer(modifier = Modifier.padding(15.dp))
        Text(
            modifier = Modifier.padding(start = 6.dp, end = 6.dp),
            text = "Welcome back to PlayZone! Enter your email address and your password to enjoy the latest features of PlayZone",
            color = Theme.colors.hintTextColor,
            textAlign = TextAlign.Center,
            fontSize = 14.sp
        )
        Spacer(modifier = Modifier.height(50.dp))
        CommonTextField(
            text = viewState.email,
            hint = "Email Address",
            enabled = !viewState.isSending,
            onValueChanged = { eventHandler(LoginEvent.EmailChanged(it)) }
        )
        Spacer(modifier = Modifier.height(24.dp))
        CommonTextField(
            text = viewState.password,
            hint = "Password",
            isSecure = viewState.passwordHidden,
            enabled = !viewState.isSending,
            trailingIcon = {
                Icon(
                    modifier = Modifier.clickable {
                        eventHandler(LoginEvent.ShowPasswordClick)
                    },
                    imageVector = if (viewState.passwordHidden) {
                        Icons.Outlined.Lock
                    } else {
                        Icons.Outlined.Clear
                    },
                    contentDescription = "",
                    tint = Theme.colors.hintTextColor
                )
            },
            onValueChanged = { eventHandler(LoginEvent.PasswordChanged(it)) }
        )
        Spacer(modifier = Modifier.height(30.dp))
        Box(modifier = Modifier.fillMaxWidth(), contentAlignment = Alignment.CenterEnd) {
            Text(
                modifier = Modifier
                    .clickable {
                        eventHandler.invoke(LoginEvent.ForgotClick)
                    },
                text = "Forgot password",
                color = Theme.colors.primaryAction,
                fontSize = 12.sp,
                textDecoration = TextDecoration.Underline
            )
        }
        Spacer(modifier = Modifier.height(40.dp))
        Button(
            modifier = Modifier
                .fillMaxWidth()
                .height(56.dp),
            enabled = !viewState.isSending,
            colors = ButtonDefaults.buttonColors(
                backgroundColor = Theme.colors.primaryAction
            ),
            shape = Theme.shapes.default,
            onClick = {
                eventHandler(LoginEvent.LoginClick)
            }) {
            Text(
                text = "Login now",
                color = Theme.colors.primaryTextColor,
                fontSize = 16.sp,
                fontWeight = FontWeight.Bold
            )
        }
    }
}